packages/router/src/lib/operations/metaTx.ts

```diff
@@ -0,0 +1,267 @@
+    // router shouldn't listen to metaTx request in first place if it's not allowed to relay
+    throw new NotAllowedFulfillRelay(chainId, {
+      methodContext,
+      requestContext,
+      data,
+      input,
+    });
+  }
+
+  // receiver fulfill, done directly on contract using user's sig broadcast
+  if (type === MetaTxTypes.Fulfill) {
+    const { txData, signature, relayerFee, callData } = data as MetaTxFulfillPayload;
+    // Send to tx service
+    logger.info("Sending fulfill tx", requestContext, methodContext, { signature });
+
+    let outputDecimals = chainData.get(txData.receivingChainId.toString())?.assetId[txData.receivingAssetId]?.decimals;
+    if (!outputDecimals) {
+      outputDecimals = await txService.getDecimalsForAsset(txData.receivingChainId, txData.receivingAssetId);
+    }
+    logger.info("Got output decimals", requestContext, methodContext, { outputDecimals });
+    const expectedFulfillFee = await calculateGasFeeInReceivingTokenForFulfill(
+      txData.receivingAssetId,
+      txData.receivingChainId,
+      outputDecimals,
+      requestContext,
+    );
+    logger.info("Expected Fulfill fee in router side", requestContext, methodContext, {
+      expectedFulfillFee: expectedFulfillFee.toString(),
+    });
+
+    const recvAmountLowerBound = expectedFulfillFee.mul(100 - relayerFeeLowerBound).div(100);
+    if (BigNumber.from(data.relayerFee).lt(recvAmountLowerBound)) {
+      throw new NotEnoughRelayerFee(chainId, {
+        methodContext,
+        requestContext,
+        relayerFee: data.relayerFee,
+        recvAmountLowerBound: recvAmountLowerBound.toString(),
+        input,
+        type,
+      });
+    }
+
+    receipt = await contractWriter.fulfillTransactionManager(
+      chainId,
+      {
+        txData,
+        signature: signature,
+        relayerFee: relayerFee,
+        callData: callData,
+      },
+      requestContext,
+    );
+
+    logger.info("Method complete", requestContext, methodContext, { transactionHash: receipt.transactionHash });
+  } else {
+    // router contract methods
+    const relayerFeeAsset = config.chainConfig[chainId].routerContractRelayerAsset ?? constants.AddressZero;
+    let relayerFeeAssetDecimal =
+      relayerFeeAsset === constants.AddressZero
+        ? 18
+        : chainData.get(chainId.toString())?.assetId[relayerFeeAsset]?.decimals;
+    if (!relayerFeeAssetDecimal) {
+      relayerFeeAssetDecimal = await txService.getDecimalsForAsset(chainId, relayerFeeAsset);
+    }
+
+    if (type === MetaTxTypes.RouterContractPrepare) {
+      const {
+        params: { txData, amount, bidSignature, encodedBid, encryptedCallData, expiry },
+        signature,
+        relayerFee,
+      } = data as MetaTxRouterContractPreparePayload;
+
+      const routerRelayerFee = await calculateGasFee(
+        chainId,
+        relayerFeeAsset,
+        relayerFeeAssetDecimal,
+        "prepare",
+        requestContext,
+        methodContext,
+      );
```

Comment:
all of these gas fees need to have the adjustments for the costs of calling router prepare vs direct contract prepare